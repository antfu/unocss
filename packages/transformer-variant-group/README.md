# @unocss/transformer-variant-group

Enables the [variant group feature of Windi CSS](https://windicss.org/features/variant-groups.html) for UnoCSS.

## Install

```bash
npm i -D @unocss/transformer-variant-group
```

```ts
import Unocss from 'unocss/vite'
import transformerVariantGroup from '@unocss/transformer-variant-group'

Unocss({
  transformers: [
    transformerVariantGroup(),
  ],
})
```

## Usage

```html
<div class="hover:(bg-gray-400 font-medium) font-(light mono)"/>
```

Will be transformed to:

```html
<div class="hover:bg-gray-400 hover:font-medium font-light font-mono"/>
``` 

## License

MIT License Â© 2021-PRESENT [Anthony Fu](https://github.com/antfu)
