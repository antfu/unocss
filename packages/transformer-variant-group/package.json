{
  "name": "@unocss/transformer-variant-group",
  "version": "0.26.4",
  "description": "Variant group transformer for UnoCSS",
  "keywords": [
    "unocss",
    "unocss-transformer"
  ],
  "homepage": "https://github.com/unocss/unocss/tree/main/packages/transformer-variant-group#readme",
  "bugs": {
    "url": "https://github.com/unocss/unocss/issues"
  },
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/unocss/unocss.git",
    "directory": "packages/transformer-variant-group"
  },
  "funding": "https://github.com/sponsors/antfu",
  "author": "Anthony Fu <anthonyfu117@hotmail.com>",
  "sideEffects": false,
  "files": [
    "dist"
  ],
  "exports": {
    ".": {
      "require": "./dist/index.cjs",
      "import": "./dist/index.mjs"
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "scripts": {
    "build": "unbuild",
    "stub": "unbuild --stub"
  },
  "dependencies": {
    "@unocss/core": "workspace:*"
  },
  "devDependencies": {
    "magic-string": "^0.25.7"
  }
}
