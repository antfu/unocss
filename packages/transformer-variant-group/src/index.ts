import type { SourceCodeTransformer } from '@unocss/core'
import { expandVariantGroup, extractQuoted } from '@unocss/core'

export default function transformerVariantGroup(): SourceCodeTransformer {
  return {
    name: 'variant-group',
    enforce: 'pre',
    transform(s) {
      extractQuoted(
        s.toString(),
        {
          details: true,
          templateStaticOnly: true,
          deep: true,
        },
      )
        .filter(({ value: { length } }) => length)
        .forEach(({ value, range }) => s.overwrite(...range, expandVariantGroup(value)))
    },
  }
}
