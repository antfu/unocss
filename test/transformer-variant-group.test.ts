import { describe, expect, test } from 'vitest'
import { transformVariantGroups } from '@unocss/transformer-variant-group'
import MagicString from 'magic-string'

describe('transformer-variant-group', () => {
  async function transform(code: string) {
    const s = new MagicString(code)
    transformVariantGroups(s)
    return s.toString()
  }

  test('basic', async() => {
    const cases = [
      'bg-white font-light sm:hover:(bg-gray-100 font-medium)',
      'lt-sm:hover:(p-1 p-2)',
      'sm:(p-1 p-2)',
      'dark:lg:(p-1 p-2)',
      'at-lg:(p-1 p-2)',
      'md:(w-40vw pr-4.5rem)',
      'lt-md:(grid grid-cols-[1fr,50%])',
      '!hover:(m-2 p-2)',
      'hover:(!m-2 p-2)',
      'md:(w-1/2 h-[calc(100%-4rem)])',
      'hover:(\n!m-2 \np-2\n)',
    ]

    for (const c of cases) {
      const result = await transform(c)
      expect(result).toMatchSnapshot(`"${c}"`)
    }
  })
})
