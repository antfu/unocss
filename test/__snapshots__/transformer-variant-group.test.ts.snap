// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformer-variant-group > basic > "!hover:(m-2 p-2)" 1`] = `"!hover:m-2 !hover:p-2"`;

exports[`transformer-variant-group > basic > "[&]:(w-4 h-4) [&]:(w-4 h-4)" 1`] = `"[&]:w-4 [&]:h-4 [&]:w-4 [&]:h-4"`;

exports[`transformer-variant-group > basic > "at-lg:(p-1 p-2)" 1`] = `"at-lg:p-1 at-lg:p-2"`;

exports[`transformer-variant-group > basic > "bg-white font-light sm:hover:(bg-gray-100 font-medium)" 1`] = `"bg-white font-light sm:hover:bg-gray-100 sm:hover:font-medium"`;

exports[`transformer-variant-group > basic > "dark:lg:(p-1 p-2)" 1`] = `"dark:lg:p-1 dark:lg:p-2"`;

exports[`transformer-variant-group > basic > "hover:(
!m-2 
p-2
)" 1`] = `"!hover:m-2 hover:p-2"`;

exports[`transformer-variant-group > basic > "hover:(!m-2 p-2)" 1`] = `"!hover:m-2 hover:p-2"`;

exports[`transformer-variant-group > basic > "lt-md:(grid grid-cols-[1fr,50%])" 1`] = `"lt-md:grid lt-md:grid-cols-[1fr,50%]"`;

exports[`transformer-variant-group > basic > "lt-sm:hover:(p-1 p-2)" 1`] = `"lt-sm:hover:p-1 lt-sm:hover:p-2"`;

exports[`transformer-variant-group > basic > "md:(w-1/2 h-[calc(100%-4rem)])" 1`] = `"md:w-1/2 md:h-[calc(100%-4rem)]"`;

exports[`transformer-variant-group > basic > "md:(w-40vw pr-4.5rem)" 1`] = `"md:w-40vw md:pr-4.5rem"`;

exports[`transformer-variant-group > basic > "sm:(p-1 p-2)" 1`] = `"sm:p-1 sm:p-2"`;
